### PR TITLE
ARROW-7800 [Python] implement iter_batches() method for ParquetFile and ParquetReader

### PR DESCRIPTION
**Pull request by ghost**
Implements an `iter_batches()` method for `ParquetFile` and `ParquetReader`. This is an attempt to complete the work started in #6386 by @rollokb.

I wasn't sure how to set the `batch_size` parameter in `ArrowReaderProperties` after the init, so I ended up expanding the C++ API so that `GetRecordBatchReader` took a `batch_size` parameter. I know nearly nothing about C++ and Cython, so if there's a better way I'd love to learn. But I saw [Wes' comment in the PR creating those methods suggesting you might want this](https://github.com/apache/arrow/pull/4304#pullrequestreview-253163586), so I thought there's a chance I'm not overstepping too much.

### CODE REVIEW

File: python/pyarrow/_parquet.pyx
Comment (fsaintjacques) [inline python/pyarrow/_parquet.pyx:1087]: Ditto, I'd add a method `def set_batch_size(self, int64_t batch_size)`. Why do you need the `batch_size` and `use_threads` if they're already property of the class and the caller can change them via setters of the instance?

**Review by fsaintjacques — CHANGES_REQUESTED** [review]

File: cpp/src/parquet/arrow/reader.cc
Comment (wjones1) [reply cpp/src/parquet/arrow/reader.cc:263]: Thanks for the feedback. I think I understand now the pattern going on with `set_use_threads()`. It's a method on the reader that [calls the method of the same name it's `ArrowReaderProperties`](https://github.com/apache/arrow/blob/master/cpp/src/parquet/arrow/reader.cc#L279-L281).

I will implement the same thing for [the `set_batch_size()`](https://github.com/apache/arrow/blob/master/cpp/src/parquet/properties.h#L564) method.

File: python/pyarrow/_parquet.pxd
Comment (wjones1) [reply python/pyarrow/_parquet.pxd:337]: 👍

File: python/pyarrow/_parquet.pyx
Comment (wjones1) [reply python/pyarrow/_parquet.pyx:1087]: I have those parameters because I was following the pattern set by the [`read_row_group()` and `read_row_groups()`](https://github.com/apache/arrow/blob/66a18e14f64359b73bd8ed5625c3092cd369af44/python/pyarrow/_parquet.pyx#L1131-L1136) methods in the same file, which both have a `use_threads` parameters. Same with the `read_all()` method.

File: python/pyarrow/tests/test_parquet.py
Comment (rollokb) [inline python/pyarrow/tests/test_parquet.py:207]: @wjones1  thank you very much for taking over my PR. Unfortunately one of the reasons I abandoned this (and just built a wheel for my own company's project), was not knowing what to do about the problem this test highlights.

The problem with dictionary arrays is far beyond my skill. I was wondering what someone with more architectural vision thinks of this issue.

Comment (wjones1) [reply python/pyarrow/tests/test_parquet.py:207]: Strangely, after I merged the latest changes from master, I am no longer seeing this issue with dictionary arrays. I definitely saw it in the original fork, so I think it may have actually been fixed (though not sure where).

I've removed the dictionary array correction in the test and hopefully the CI should confirm what I am seeing.

**Comment by wjones1** [comment]
Two failing checks right now.

For the linting one, it seems to be alarmed by some Rust code that I didn't touch. Am I missing something in that output?

For the Python tests, getting [this error](https://github.com/apache/arrow/pull/6979/checks?check_run_id=618464479#step:7:2128):
```
pyarrow.lib.ArrowNotImplementedError: This class cannot yet iterate chunked arrays
```
When I run the same test locally, it succeeds.

**Comment by wjones1** [comment]
I found the cause of the test failure: If the `batch_size` isn't aligned with the `chunk_size`, categorical columns will fail with the error:
```
pyarrow.lib.ArrowNotImplementedError: This class cannot yet iterate chunked arrays
```

I think this means categorical columns/DictionaryArray columns aren't supported by this method for now, except if you are able to align the `batch_size` with `chunk_size`. 

Is it possible or even common that `chunk_size` might be variable within a file?

(The reason we were seeing the error in Python 3.5 and not in later Python versions is I was selecting a subset of columns using indices, and the ordering of columns changed between Python versions. I think because of the change in dictionary ordering in 3.6+. I've instead moved to have the offending test run on all columns.)

File: python/pyarrow/_parquet.pyx
Comment (fsaintjacques) [reply python/pyarrow/_parquet.pyx:1084]: Noted for consistency. I find it un-intuitive that calling this method changes the batch size (or threading) for the following callers. Since this is not introduced by this PR, I won't make this a blocking issue.

I'll leave @jorisvandenbossche decide if we should have a followup to fix this anti-pattern.

File: python/pyarrow/_parquet.pyx
Comment (fsaintjacques) [inline python/pyarrow/_parquet.pyx:1125]:
```suggestion
            yield pyarrow_wrap_batch(record_batch)
```

File: python/pyarrow/parquet.py
Comment (fsaintjacques) [inline python/pyarrow/parquet.py:313]: In the other method `scan_contents`, batch_size is optional.

**Comment by jorisvandenbossche** [comment]
Same question as in the other PR: does setting the batch size also influence existing methods like `read` or `read_row_group` ? Should we add that keyword there as well?

**Comment by sonthonaxrk** [comment]
@jorisvandenbossche yes it would affect how `read` works as it changes the reader properties.

I'm not too fussed about adding it into read as you can use `set_batch_size` to set the internals.

**Comment by sonthonaxrk** [comment]
@wjones1 could I have write access to your branch so I can finish the changes without this going stale again?

**Comment by wesm** [comment]
@sonthonaxrk I would recommend opening a new PR

File: python/pyarrow/parquet.py
Comment (wjones1) [reply python/pyarrow/parquet.py]: Agreed. I added the same default value for `iter_batches()`

File: python/pyarrow/_parquet.pyx
Comment (wjones1) [reply python/pyarrow/_parquet.pyx]: 👍

**Comment by wjones1** [comment]
Apologies, I've been away for a bit. I thought I had invited @sonthonaxrk as a collaborator on my fork, but perhaps that did go through.

Addressed the minor feedback.

Also retested the issue I was having where tests would fail if the `batch_size` and `chunk_size` weren't aligned, and I can't reproduce that error any more. Hopefully that means it was fixed and I'm not forgetting something important.

**Comment by wjones1** [comment]
RE: @jorisvandenbossche 
> Same question as in the other PR: does setting the batch size also influence existing methods like `read` or `read_row_group` ? Should we add that keyword there as well?

My one hesitation to adding them is that it's not clear to me what the effect would be on the execution. For `iter_batches()` the effect of `batch_size` is quite obvious, but I'm not sure about these other methods. 

After quick search of the Apache Arrow docs the only explanation I saw on the batch size parameter was this:

>  [The maximum row count for scanned record batches. If scanned record batches are overflowing memory then this method can be called to reduce their size.](https://arrow.apache.org/docs/python/generated/pyarrow.dataset.Scanner.html?highlight=batch_size)



If I can find a good explanation for it or if you have one, I'd be happy to add the `batch_size` parameter to the `read()` and `read_row_group()` methods.

**Comment by wjones1** [comment]
Actually @jorisvandenbossche, I agree we should probably just add in the batch_size argument (with a sensible default) to those other methods. Took me a while to understand what you were getting at, but by providing those parameters we avoid the issue where `iter_batches()` calls could mess with the underlying parameters when calling the other methods.

I will implement those soon. I would especially welcome input on what the docstring should be for the `batch_size` parameter for those other methods (what is the user-facing impact of changing it?) and what the default value should be.

**Comment by wjones1** [comment]
Looking at the code, no longer think this `batch_size` parameter actually would affect those other read methods.

There are a few different "batch_size" parameters floating around the `reader.cc`, but there's only one reference to the one in `reader_properties_` (`ArrowReaderProperties`): 
https://github.com/apache/arrow/blob/edf24290046d95967d620104c5238f30ff032b6d/cpp/src/parquet/arrow/reader.cc#L797-L799

As far as I can tell, that's exclusively on the code path for the RecordBatchReader, and not the other readers. So I don't think we need to add the parameter to those other methods.

**Comment by jorisvandenbossche** [comment]
@wjones1 sorry for the slow follow-up. Could you rebase the PR?

**Review by wesm — APPROVED** [review]
+1. I rebased. Will merge this and I suspect that this interface can be refined in follow up PRs

**Comment by wesm** [comment]
Ah, well the tests are failing after rebasing so someone needs to take a closer look

**Comment by wjones1** [comment]
So it appears there were changes to the underlying implementation of RecordBatchReader. Prior to these changes, it would yield record batches with the exact batch size (if possible). So for a `batch_size` of 900 on a file written with `chunk_size` of 1,000, it would yield batches of: 900, 900, 900, 900, ... rows. Now it yields slices that are aligned with the rowgroups, so the same parameters would yield batches with rowcounts: 900, 100, 900, 100, and so on.

I'm not 100% sure if we care about the exact number of rows returned, but for now I'm leaning towards yes. Open to feedback on that. In the meantime, I will push changes soon that will stitch together the batches to yield consistent rowcounts.

A cool side affect of these changes is is that it gets around the bug I mentioned earlier that would have blocked support for categorical columns in this method.

**Comment by wjones1** [comment]
On second thought, I think if users was consistent batch_sizes they can probably add that functionality in a wrapping generator.

I have adjusted my tests to the new behavior and edited the docstring to indicate that the `batch_size` parameter is a maximum.

**Comment by wjones1** [comment]
I have the unittests passing locally but they seem to be failing in CI. I probably need to rebase again and test. Will do so when I get some time this weekend.

**Comment by emkornfield** [comment]
@wjones1 did you have a chance to figure out CI issues?

**Comment by jorisvandenbossche** [comment]
@wjones1 I tested this branch locally, and can actually reproduce the errors. 

I don't see what you described above (https://github.com/apache/arrow/pull/6979#issuecomment-659143356) about batch_size now being a max number (the last batch of a single row group thus being potentially smaller). 

When debugging the test for batch_size of 300 with chunk_size of 1000, I don't see the 4th batch having 100 rows (rows 900-1000), but actually also 300 (rows 900-1200).

File: python/pyarrow/parquet.py
Comment (arw2019) [inline python/pyarrow/parquet.py:326]: Should be `Returns`

File: python/pyarrow/parquet.py
Comment (arw2019) [inline python/pyarrow/parquet.py:329]: `Contents`

File: python/pyarrow/parquet.py
Comment (wjones1) [reply python/pyarrow/parquet.py]: Got it. Also realized that technically it's yielding RecordBatches, not Tables. So changing to `iterator of pyarrow.RecordBatch`.

**Comment by wjones1** [comment]
I'm back on this for the weekend and will be back as needed the week after next.

@jorisvandenbossche I can confirm that once I merge in the latest changes from apache master, I am getting the batches to be the expected size (and spanning across the parquet chunks). I have updated the `test_iter_batches_columns_reader` unit test accordingly.

However, I have found that if the columns selected include a categorical column, then it reverts back to the behavior I was seeing before. You should be able to reproduce this by editing line 178 of the parquet tests to include `categorical=True`, and will find that the `test test_iter_batches_columns_reader` will then fail. I'm fine leaving this fix for a future story.

**Comment by wjones1** [comment]
Got the actions to rerun, but some are still failing. As far as I can tell, these failures aren't due to the changes in this PR.

**Comment by emkornfield** [comment]
@wjones1 I think the ming/windows ones are ignorable.

Lint one is real:
```
INFO:archery:Running Python formatter (autopep8)
INFO:archery:Running Python linter (flake8)
/arrow/python/pyarrow/tests/test_parquet.py:25:1: F401 'os.stat_result' imported but unused
/arrow/python/pyarrow/tests/test_parquet.py:189:80: E501 line too long (86 > 79 characters)
```

**Comment by wjones1** [comment]
@emkornfield  Thanks for catching that. I've fixed the formatting issues. Looks like it's not just the windows R checks that are failing.

**Comment by nealrichardson** [comment]
The R Windows failures are spurious, you can ignore.

**Comment by emkornfield** [comment]
@wesm does your +1 still hold?  Can this be merged now?

File: python/pyarrow/_parquet.pyx
Comment (wesm) [inline python/pyarrow/_parquet.pyx:1111]: These function calls do non-trivial computing -- need to release the GIL

File: python/pyarrow/parquet.py
Comment (wesm) [inline python/pyarrow/parquet.py:341]: Doesn't `iter_batches` return a generator, so you can just `return self.reader.iter_batches(...)`?

**Review by wesm — CHANGES_REQUESTED** [review]
I noted one significant issue (a missing GIL release) and another minor one that would be good to fix too

**Comment by wjones127** [comment]
Update: I have just rebased. Unit tests still looking good, and I think should be still ready for merge.

Note that I merged my old Github user into this one, so that's why this appears to be opened by "Ghost".

There is one CI failure for "C++ / AMD64 Windows 2019 C++", but I can't figure out where the error is in there. It does seem like I am not the only one getting that failure (see https://github.com/apache/arrow/pull/8834/checks?check_run_id=1500735491), so I'm skeptical that it's due to my changes.

**Comment by wesm** [comment]
cc @jorisvandenbossche

**Comment by jorisvandenbossche** [comment]
(@wjones127 it's on my to do list to give this another look next week when back from the holidays)

**Comment by jorisvandenbossche** [comment]
Trying this out locally, I see the following strange behaviour:

```
In [90]: size = 300

In [91]: table = pa.table({'str': [str(x) for x in range(size)]})

In [92]: pq.write_table(table, "test.parquet", row_group_size=100)

In [93]: f = pq.ParquetFile("test.parquet")

In [94]: [b.num_rows for b in f.iter_batches(batch_size=80)]
Out[94]: [80, 80, 80, 60]

In [95]: table = pa.table({'str': pd.Categorical([str(x) for x in range(size)])})

In [96]: pq.write_table(table, "test.parquet", row_group_size=100)

In [97]: f = pq.ParquetFile("test.parquet")

In [98]: [b.num_rows for b in f.iter_batches(batch_size=80)]
Out[98]: [80, 20, 60, 40, 40, 60]
```

So by having a dictionary type, the size of the batches is rather strange. Now, it was already discussed above that with categorical data, it was behaving differently (and that's also the reason the tests don't include categorical for now). So since you are mostly exposing the existing feature in python, and not actually implementing it, I suppose this is an existing bug in the batch_size handling of the parquet C++ record batch reader, and it can be seen as a bug to report/fix separately?

> I can confirm that once I merge in the latest changes from apache master, I am getting the batches to be the expected size (and spanning across the parquet chunks). I have updated the test_iter_batches_columns_reader unit test accordingly.

That's again something tied to the C++ implementation, but I would actually not expect the baches to cross different row groups (so for example for a row group size of 1000, total size of 2000 and batch size of 300, I would expect batches of [300, 300, 300, 100, 300, 300, 300, 100] instead of [300, 300, 300, 300, 300, 300, 200]. 
But so maybe something to open a JIRA for to follow-up on.

> Looking at the code, no longer think this `batch_size` parameter actually would affect those other read methods.
> 
> There are a few different "batch_size" parameters floating around the `reader.cc`, but there's only one reference to the one in `reader_properties_` (`ArrowReaderProperties`):
> ...
> As far as I can tell, that's exclusively on the code path for the RecordBatchReader, and not the other readers. So I don't think we need to add the parameter to those other methods.

That indeed seems to be the case

File: python/pyarrow/parquet.py
Comment (jorisvandenbossche) [inline python/pyarrow/parquet.py:331]:
```suggestion
            smaller if there aren't enough rows in the file.
```

? (given that it currently returns row across row groups)

File: python/pyarrow/parquet.py
Comment (jorisvandenbossche) [inline python/pyarrow/parquet.py:342]:
```suggestion
        columns: list
            If not None, only these columns will be read from the file. A
            column name may be a prefix of a nested field, e.g. 'a' will select
            'a.b', 'a.c', and 'a.d.e'.
        use_threads : boolean, default True
            Perform multi-threaded column reads.
        use_pandas_metadata : boolean, default False
            If True and file has custom pandas schema metadata, ensure that
            index columns are also loaded.
```

(minor nitpick on punctuation)

**Review by jorisvandenbossche — APPROVED** [review]
two small nitpicks on the docstrings, you can also just commit them here directly on github

**Comment by jorisvandenbossche** [comment]
BTW, the Dataset API also gives a way to get an iterator over record batches (with `Dataset.to_batches()`). The strange thing is that this seems to have another logic of how many rows are included in each batch when crossing row groups, while in the end it is also using `GetRecordBatchReader` with `batch_size` set in the reader properties:

```
In [116]: table = pa.table({'str': [str(x) for x in range(size)], 'str_cat': pd.Categorical([str(x) for x in range(size)])})

In [117]: pq.write_table(table, "test.parquet", row_group_size=100)

In [118]: f = pq.ParquetFile("test.parquet")

In [119]: [b.num_rows for b in f.iter_batches(batch_size=80)]
Out[119]: [80, 20, 60, 40, 40, 60]

In [120]: [b.num_rows for b in ds.dataset("test.parquet").to_batches()]
Out[120]: [100, 100, 100]

In [121]: [b.num_rows for b in ds.dataset("test.parquet").to_batches(batch_size=80)]
Out[121]: [80, 20, 80, 20, 80, 20]
```

**Comment by jorisvandenbossche** [comment]
> The strange thing is that this seems to have another logic of how many rows are included in each batch when crossing row groups, while in the end it is also using `GetRecordBatchReader` with `batch_size` set in the reader properties:

Ah, that's because the RecordBatchReader gets constructed *for each* row group, so that way never crossing row group boundaries

**Comment by jorisvandenbossche** [comment]
Did another rebase in the hope that that fixes the CI failures

**Comment by jorisvandenbossche** [comment]
The Mac failures are unrelated (install issue, failing elsewhere as well). Since other python builds are passing, and I don't expect this change to be especially platform-sensitive, going to merge this.

**Comment by jorisvandenbossche** [comment]
Thanks again @wjones127. I opened https://issues.apache.org/jira/browse/ARROW-11142 about the inconsistencies in batch sizing.

**Pull request closed without merging**
